### PR TITLE
balance proposal

### PR DESCRIPTION
**Pull request by philippe44**
Here is what 'balance' could look like if we think it is worthwhile. Please do not merge it at this point, it's just an idea. I've made it a player capability so that not all squeezeplay have the slider option.

### CODE REVIEW

File: HTML/EN/settings/player/audio.html [hunk 1]
```diff
@@ -427,4 +427,10 @@
 		[% END %]
 	[% END %]
 
+	[% IF prefs.exists('pref_balance') %]
+		[% WRAPPER setting title="SETUP_BALANCE" desc="SETUP_BALANCE_DESC" %]
+			<input type="text" class="stdedit sliderInput_-25_25" name="pref_balance" id="pref_balance" value="[% prefs.pref_balance %]" size="5">
+		[% END %]
+	[% END %]
+
```
Comment (michaelherger) [inline HTML/EN/settings/player/audio.html]: Please move it up to bass/treble.

File: Slim/Player/SqueezePlay.pm [hunk 2]
```diff
@@ -89,6 +91,7 @@ my %CapabilitiesMap = (
 	Spdirect                => 'spDirectHandlers',
 	Proxy                   => 'proxyAddress',
 	CanHTTPS                => '_canHTTPS',
+	Balance					=> 'balance',
```
Comment (michaelherger) [inline Slim/Player/SqueezePlay.pm]: Whitespace ;-)

File: Slim/Player/Squeezebox2.pm [hunk 3]
```diff
@@ -296,15 +297,18 @@ sub volume {
 		my $preamp = 255 - int( 2 * ( $prefs->client($client)->get('preampVolumeControl') || 0 ) );
 
 		my $data;
+		my $balance = $prefs->client($client)->get('balance') || 0;
+		my $left = $balance > 0 ? (25 - $balance) / $balance : 1;
+		my $right = $balance < 0 ? (25 + $balance) / -$balance : 1;
```
Comment (michaelherger) [inline Slim/Player/Squeezebox2.pm:304]: Don't we risk to pull one channel to eleven? Should we cap the channel volume, or does the player do this?

**Review by michaelherger — COMMENTED** [review]
I'll have to let more knowledgeable people review the technical part of the implementation... 

Searching around LMS code I find `balanceBar()` - which never seems to be called. And you mentioned SB1 having balance controls? I'll have to double check. And I'm wondering why they would have removed it for SB2.

As this is all done server side I would assume I could test this already?

File: HTML/EN/settings/player/audio.html (hunk 1 above)
Comment (philippe44) [reply HTML/EN/settings/player/audio.html]: done

File: Slim/Player/Squeezebox2.pm (hunk 3 above)
Comment (philippe44) [reply Slim/Player/Squeezebox2.pm:304]: I'm a bit confused here, why eleven? $balance is -25..25 and $left & $right are supposed to be ratio where one will be set to 1 and the other within [0..1[ that would be a multiplier to gain. Very basic...

File: Slim/Player/SqueezePlay.pm (hunk 2 above)
Comment (philippe44) [reply Slim/Player/SqueezePlay.pm]: I need to have at least one of these :-)

File: Slim/Player/Squeezebox2.pm (hunk 3 above)
Comment (michaelherger) [reply Slim/Player/Squeezebox2.pm:304]: https://youtu.be/4xgx4k83zzc 😁 

We should have some `setValidate()` in place to make sure it doesn't go beyond the expected +/-25 (see line 57). The slider is only rendered in the default skin, but not in Classic and more importantly Material.

**Review by michaelherger — CHANGES_REQUESTED** [review]
Please add a `$prefs->setValidate()` statement to make sure we don't go beyond the expected values.

**Comment by philippe44** [comment]
> Please add a `$prefs->setValidate()` statement to make sure we don't go beyond the expected values.
>
Done - that reminds me that this -25/25 is hardcoded at least in 3 places which hurts my feeling :-)

File: Slim/Player/SqueezePlay.pm [hunk 4]
```diff
@@ -89,6 +91,7 @@ my %CapabilitiesMap = (
 	Spdirect                => 'spDirectHandlers',
 	Proxy                   => 'proxyAddress',
 	CanHTTPS                => '_canHTTPS',
+	Balance                 => 'balance',
```
Comment (mherger) [inline Slim/Player/SqueezePlay.pm:94]: Just wanted to give this a try... but which player would support balance?

Comment (philippe44) [reply Slim/Player/SqueezePlay.pm:94]: I just pushed an update to squeezelite-esp32. I need to propose something to squeezelite, but it's just adding one item in the HELO string "Balance=1"

**Comment by mherger** [comment]
Ok, this seems to be working just fine with the SqueezeAMP. Now... it requires me to restart playback in order to hear a change in audio. Would there be any way to make this more "real time"?

**Comment by mherger** [comment]
Thanks!

**Pull request merged**